modules/3d/src/octree.cpp

```diff
@@ -0,0 +1,507 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html
+
+#include "precomp.hpp"
+#include "octree.hpp"
+
+namespace cv{
+
+OctreeNode::OctreeNode():children(8, nullptr), depth(0), size(0), origin(0,0,0), parentIndex(-1)
+{
+}
+
+OctreeNode::OctreeNode(int _depth, double _size, const Point3f& _origin, int _parentIndex):children(8), depth(_depth), size(_size), origin(_origin), parentIndex(_parentIndex)
+{
+}
+
+bool OctreeNode::isPointInBound(const Point3f& _point)
+{
+    return isPointInBound(_point, origin, size);
+}
+
+bool OctreeNode::isPointInBound(const Point3f& _point, const Point3f& _origin, double _size)
+{
+    if((_point.x + epsilon > _origin.x && _point.y + epsilon > _origin.y && _point.z + epsilon > _origin.z) &&
+       (_point.x < _origin.x + _size + epsilon && _point.y < _origin.y + _size + epsilon && _point.z < _origin.z + _size + epsilon))
```

Comment:
In general, it is not correct to use the same epsilon value.

`std::numeric_limits<float>::epsilon()` can be used safely to compare values near 1.0 (0.5 ... 1.0 in the worst case).

It is not applicable to compare large (100000.0) or small (1e-6) numbers.

Probably we should adjust epsilon values first (still approximation):
```
float epsX = std::numeric_limits<float>::epsilon() * std::max(std::abs(_point.x), std::abs(_origin.x))
float epsY = ...;
float epsZ = ...;
```

See also: https://en.wikipedia.org/wiki/Machine_epsilon

---

Using strong comparison doesn't look correct:
```.diff
-_point.x + epsilon > _origin.x
+_point.x + epsilon >= _origin.x
```

Reply:
Thanks, fixed.